src/daft-logical-plan/src/optimization/rules/reorder_joins/brute_force_join_order.rs

```diff
@@ -0,0 +1,379 @@
+
+// Takes the following arguments:
+// - elements: The list of elements to choose from.
+// - cur_idx: The current index in the list of elements to choose from.
+// - remaining: The number of elements to choose.
+// - chosen: The list of elements that have already been chosen.
+// - unchosen: The list of elements that have not yet been chosen.
+//
+// Returns a list of tuples, where each tuple contains two lists:
+// - The first list is the list of elements that have been chosen.
+// - The second list is the list of elements that have not been chosen.
+fn generate_combinations(
+    elements: &[usize],
+    cur_idx: usize,
+    remaining: usize,
+    chosen: Vec<usize>,
+    mut unchosen: Vec<usize>,
+) -> Vec<(Vec<usize>, Vec<usize>)> {
+    if remaining == 0 {
+        for &element in elements.iter().skip(cur_idx) {
+            unchosen.push(element);
+        }
+        return vec![(chosen, unchosen)];
+    }
+    if cur_idx >= elements.len() {
+        return vec![];
+    }
+    let mut chosen_clone = chosen.clone();
+    chosen_clone.push(elements[cur_idx]);
+    let mut unchosen_clone = unchosen.clone();
+    unchosen_clone.push(elements[cur_idx]);
+    let mut results =
+        generate_combinations(elements, cur_idx + 1, remaining, chosen, unchosen_clone);
+    results.extend(generate_combinations(
+        elements,
+        cur_idx + 1,
+        remaining - 1,
+        chosen_clone,
+        unchosen,
+    ));
+    results
+}
+
+impl BruteForceJoinOrderer {
+    // Enumerates all possible join orders and returns the one with the lowest summed cardinality.
+    fn find_min_cost_order(
+        graph: &JoinGraph,
+        available: Vec<usize>,
+    ) -> Option<(usize, JoinOrderTree)> {
+        if available.len() == 1 {
+            // Base case: if there is only one element, we return the cost of the relation and the join order tree.
+            let id = available[0];
+            let plan = graph
+                .adj_list
+                .id_to_plan
+                .get(&id)
+                .expect("Got non-existent ID in join graph");
+            let stats = plan.materialized_stats();
+            let cost = stats.approx_stats.num_rows;
+            return Some((cost, JoinOrderTree::Relation(id, cost)));
+        }
+        // Recursive case: we split the available elements into two groups and recursively find the minimum cost join order for each group.
+        // We only need to consider splits where the left group has at most half the number of elements as the right group, because the
+        // cardinality of the join is commutative. Determining probe/build sides happens later in the physical planner.
+        let max_left_size = available.len() / 2;
+        let mut min_cost = None;
+        let mut chosen_plan = None;
+        for left_split_size in 1..=max_left_size {
+            for (chosen, unchosen) in
+                generate_combinations(&available, 0, left_split_size, vec![], vec![])
```

Comment:
I would favor heap's algorithm that can do this in place over allocating every combination

https://en.wikipedia.org/wiki/Heap%27s_algorithm

Reply:
Took inspiration from the algorithm, but ultimately decided to adapt it instead. The algorithm produces all n! permutations, when we really just need the (n choose k) combinations, which is a subset of n!.

I did remove all the heap allocations and now rely solely on swapping and backtracking.